notifier/service/service.go

```diff
@@ -0,0 +1,223 @@
+		p := notifier.NewProcessor(
+			i,
+			distLock,
+			opts.Indexer,
+			opts.Matcher,
+			store,
+		)
+		p.Process(ctx, c)
+	}
+
+	// kick off configured deliverer type
+	switch {
+	case opts.Webhook != nil:
+		if err := webhookDeliveries(ctx, opts, lockPool, store); err != nil {
+			return nil, err
+		}
+	case opts.AMQP != nil:
+		if err := amqpDeliveries(ctx, opts, lockPool, store); err != nil {
+			return nil, err
+		}
+	}
+
+	return &service{
+		store: store,
+	}, nil
+}
+
+func storeInit(ctx context.Context, opts Opts) (*postgres.Store, *sqlx.DB, error) {
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "notifier/service/storeInit").
+		Logger()
+	ctx = log.WithContext(ctx)
+
+	cfg, err := pgxpool.ParseConfig(opts.ConnString)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to parse ConnString: %v", err)
+	}
+	cfg.MaxConns = 30
+	pool, err := pgxpool.ConnectConfig(ctx, cfg)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create ConnPool: %v", err)
+	}
+
+	lockPool, err := sqlx.Connect("pgx", opts.ConnString)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create lock ConnPool: %v", err)
+	}
+
+	// do migrations if requested
+	if opts.Migrations {
+		log.Info().Msg("performing notifier migrations")
+		migrator := migrate.NewPostgresMigrator(lockPool.DB)
+		migrator.Table = migrations.MigrationTable
+		err := migrator.Exec(migrate.Up, migrations.Migrations...)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to perform migrations: %w", err)
+		}
+	}
+
+	log.Info().Msg("intializing notifier store")
+	store := postgres.NewStore(pool)
+	return store, lockPool, nil
+}
+
+func webhookDeliveries(ctx context.Context, opts Opts, lockPool *sqlx.DB, store notifier.Store) error {
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "notifier/service/webhookInit").
+		Logger()
+	ctx = log.WithContext(ctx)
+	log.Info().Int("count", deliveries).Msg("intializing webhook deliverers")
+
+	ds := make([]*notifier.Delivery, 0, deliveries)
+	for i := 0; i < deliveries; i++ {
+		distLock := pgdl.NewLock(lockPool, 0)
+		wh, err := webhook.New(*opts.Webhook, nil)
+		if err != nil {
+			return fmt.Errorf("failed to create webhook deliverer: %v", err)
+		}
+		delivery := notifier.NewDelivery(i, wh, opts.DeliveryInterval, store, distLock)
+		ds = append(ds, delivery)
+	}
+	for _, d := range ds {
+		d.Deliver(ctx)
+	}
+	return nil
+}
+
+func amqpDeliveries(ctx context.Context, opts Opts, lockPool *sqlx.DB, store notifier.Store) error {
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "notifier/service/amqpInit").
+		Logger()
+	ctx = log.WithContext(ctx)
+
+	err := opts.AMQP.Validate()
+	if err != nil {
+		return fmt.Errorf("amqp validation failed: %v", err)
+	}
+
+	if len(opts.AMQP.URIs) == 0 {
+		log.Warn().Msg("amqp delivery was configured with no broker URIs to connect to. delivery of notifications will not occur.")
+		return nil
+	}
+
+	fo := &namqp.FailOver{
+		Config: *opts.AMQP,
```

Comment:
This has an embedded lock, it's not okay to copy around.

Reply:
Locks are gone.